tidy3d/components/data/sim_data.py

```diff
@@ -953,52 +980,256 @@ def source_spectrum_fn(freqs):
+            sim_adj_update_dict["normalize_index"] = None
+
         # set the ADJ grid spec wavelength to the original wavelength (for same meshing)
         grid_spec_original = sim_original.grid_spec
         if sim_original.sources and grid_spec_original.wavelength is None:
             wavelength_original = grid_spec_original.wavelength_from_sources(sim_original.sources)
             grid_spec_adj = grid_spec_original.updated_copy(wavelength=wavelength_original)
             sim_adj_update_dict["grid_spec"] = grid_spec_adj
 
-        return sim_original.updated_copy(**sim_adj_update_dict)
+        return sim_original.updated_copy(**sim_adj_update_dict), adjoint_source_info
 
-    def make_adjoint_sources(self, data_vjp_paths: set[tuple]) -> list[Source]:
+    def make_adjoint_sources(self, data_vjp_paths: set[tuple]) -> dict[str, Source]:
         """Generate all of the non-zero sources for the adjoint simulation given the VJP data."""
 
-        # TODO: determine if we can do multi-frequency sources
-
         # map of index into 'self.data' to the list of datasets we need adjoint sources for
         adj_src_map = defaultdict(list)
         for _, index, dataset_name in data_vjp_paths:
             adj_src_map[index].append(dataset_name)
 
-        # gather a list of adjoint sources for every monitor data in the VJP that needs one
-        sources_adj_all = []
+        # gather a dict of adjoint sources for every monitor data in the VJP that needs one
+        sources_adj_all = defaultdict(list)
         for data_index, dataset_names in adj_src_map.items():
             mnt_data = self.data[data_index]
-            sources_adj = mnt_data.make_adjoint_sources(dataset_names=dataset_names)
-            sources_adj_all += sources_adj
+            sources_adj = mnt_data.make_adjoint_sources(
+                dataset_names=dataset_names, fwidth=self.fwidth_adj
+            )
+            sources_adj_all[mnt_data.monitor.name] = sources_adj
 
         return sources_adj_all
 
+    @property
+    def fwidth_adj(self) -> float:
+        # fwidth of forward pass, try as default for adjoint
+        normalize_index_fwd = self.simulation.normalize_index or 0
+        return self.simulation.sources[normalize_index_fwd].source_time.fwidth
+
+    def process_adjoint_sources(self, adj_srcs: list[Source]) -> AdjointSourceInfo:
+        # tuple[list[Source], Union[float, xr.DataArray], bool]:
+        """Compute list of final sources along with a post run normalization for adj fields."""
+
+        # dictionary mapping unique spatial dependence of each Source to list of time-dependencies
+        json_to_sources = defaultdict(None)
+        spatial_to_src_times = defaultdict(list)
+        for src in adj_srcs:
+            src_spatial_json = src.json(exclude={"source_time"})
+            json_to_sources[src_spatial_json] = src
+            spatial_to_src_times[src_spatial_json].append(src.source_time)
+
+        num_ports = len(spatial_to_src_times)
```

Comment:
> So you are excluding the SourceTime of all adjoint sources, and counting their unique definitions after that exclusion?

That's right.

> In some cases, like when differentiating through FieldData.flux, the adjoint source would be a CustomFieldSource, right? Does this work for those too?

Good question. I'm not really sure to be honest. Right now it would treat two otherwise identical CustomFieldSource with different data as separate "ports".  I think this is the proper handling though, since the data in the source defines the spatial dependence of `J(r, ω)` and we're trying to do here is essentially separate `J(r)` and `J(ω)`, so two different datasets should be treated as different `J(r)` I think.

The only case I think this could become slightly ambiguous for the user is if they had two identical `FieldMonitor` "`a`, and `b` and then for some reason `a` and `b` had different contributions to the objective function. In that case, we could come up with a scenario where the adjoint sources for `a` and `b` could theoretically be combined into one (ie by just summing the datasets with the corresponding source time amplitudes). But this seems like such an edge case that it's probably not worrying about to me.

Reply:
just to clarify, in the case described above, we would simply require a separate adjoint simulation for `a` and `b`, but it wouldn't error.

Reply:
My worry was rather that I'm still not sure what happens if you look at the json only, i.e. if the actual data gets compared or just a dummy placeholder tag or something that would be identical for all custom sources. If it works the way you describe it sounds good.

Reply:
Yeah, that's the edge case I had in mind. Although... technically if they're identical in their definitions, the recorded datasets should also be the same?

Ah but something weird may happen where the two are used in different ways in the objective function and so the adjoint source currents are different?

Very edgy.

Reply:
well the recorded datasets would be the same, but what we get back is the VJP not the recorded datasets. so yea if the data is used in two different ways, it wouldn't show up.

I think one way to distinguish spatial dependence this could be to instead make a copy of all of the adjoint sources with the **same** source time, and then hash these. But then I dont know exactly if the datasets are included in the hash or not..